Sources/NIO/BaseSocketChannel.swift

```diff
@@ -731,27 +760,34 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
             return
         }
 
-        guard !self.lifecycleManager.isRegistered else {
+        guard !self.lifecycleManager.isRegisteredLazily else {
             promise?.fail(error: ChannelLifecycleError.inappropriateOperationForState)
             return
         }
 
-        // Was not registered yet so do it now.
-        do {
-            // We always register with interested .none and will just trigger readIfNeeded0() later to re-register if needed.
-            try self.safeRegister(interested: [.readEOF, .reset])
-            self.lifecycleManager.register(promise: promise)(self.pipeline)
-        } catch {
+        guard self.selectableEventLoop.isOpen else {
+            let error = EventLoopError.shutdown
+            self.pipeline.fireErrorCaught0(error: error)
+            self.close0(error: error, mode: .all, promise: nil)
             promise?.fail(error: error)
+            return
         }
+
+        // we can't fully register yet as epoll would give us EPOLLHUP if bind/connect wasn't called yet.
+        self.lifecycleManager.registerLazily(promise: promise)(self.pipeline)
     }
 
     public final func registerAlreadyConfigured0(promise: EventLoopPromise<Void>?) {
         assert(self.eventLoop.inEventLoop)
         assert(self.isOpen)
         assert(!self.lifecycleManager.isActive)
         register0(promise: nil)
-        becomeActive0(promise: promise)
+        if self.lifecycleManager.isRegisteredLazily {
+            try! becomeFullyRegistered0()
+            if self.lifecycleManager.isRegisteredFully {
+                becomeActive0(promise: promise)
```

Comment:
Do these callouts not get triggered by `register0`?

Reply:
@Lukasa right now `register0` only does the pre-registration. I didn't change its name as it's on `ChannelCore`. But we could obviously rename it to `preRegister0`